p2p/src/net/mock/backend.rs

```diff
@@ -208,21 +238,30 @@ where
         }
     }
 
-    /// Send request to remote peer
+    /// Sends a request to the remote peer.
     async fn send_request(
         &mut self,
-        peer_id: &MockPeerId,
+        peer_id: MockPeerId,
         request: message::Request,
     ) -> crate::Result<MockRequestId> {
         let peer = self
             .peers
-            .get_mut(peer_id)
+            .get_mut(&peer_id)
             .ok_or(P2pError::PeerError(PeerError::PeerDoesntExist))?;
 
-        let (request_id, request) = self.request_mgr.make_request(peer_id, request)?;
-
+        let (request_id, request) = self.request_mgr.make_request(request)?;
         peer.tx.send(MockEvent::SendMessage(request)).await.map_err(P2pError::from)?;
 
+        let is_inserted = self.pending_responses.insert(request_id);
+        debug_assert!(is_inserted);
+
+        let timeout = self.p2p_config.request_timeout.clone().into();
+        let sender = self.timeouts_sender.clone();
+        tokio::spawn(async move {
```

Comment:
A new task for every request sent? This seems very excessive and can easily blow up in memory in the future as a result to bad maintenance...

Why not just share an atomic of the latest time a request was sent? The way I would do this, is have an async timer in a task that's launched when the peer is connected. In that task, we own an atomic variable that contains the time of the latest request. Once a response is received, that atomic variable is reset (to zero, unfortunately using something nice like an Option would remove atomicity). 

The timer ticks every time that timeout duration passes (or twice of that, that's an OK error to accept). Once a tick happens, we check current time versus stored atomic time. If the difference is larger than timeout, we send the timeout signal.

Reply:
While I agree that spawning a new task each time is little too much, the tasks are fairly lightweight. In the documentation is stated that a task itself requires only a single allocation and 64 bytes of memory. I understand that there is an additional overhead, but spawning thousands of tasks is OK for the tokio runtime.

Anyway, I'm going to implement your approach.

Reply:
If you're looking for a good argument why this is a bad idea, just imagine someone trying to debug while all these tasks are running ;-)

In multithreaded applications, in general, we try to minimize concurrency/tasks to make reasoning about the program easier.